Plugin crashes with wrong credentials

Hi,

the plugin would always crash on me until I realized I had changed my router and did not re-create the automation user!
I had a look at the code and it seems you 're not checking that the `client` variable is valid. For example, here: https://github.com/ddelnano/terraform-provider-mikrotik/blob/c2a1c2c1e254957f8fb92e328214be2212ee7088/mikrotik/resource_dns_record.go#L48

In the above code, `c` is nil when the credentials are wrong, which leads to this stack trace:

```
Stack trace from the terraform-provider-mikrotik_v0.9.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x626307]

goroutine 70 [running]:
github.com/go-routeros/routeros.(*Client).RunArgs(0x0, {0xc000494600, 0x2, 0xc0004ad790})
        github.com/go-routeros/routeros@v0.0.0-20210123142807-2a44d57c6730/run.go:21 +0x27
github.com/ddelnano/terraform-provider-mikrotik/client.Mikrotik.FindDnsRecord({{0xc0004ad4a0, 0x10}, {0xc0004ad4d0, 0x9}, {0xc0004716f8, 0x18}, 0x0, {0x0, 0x0}, 0x0, ...}, ...)
        github.com/ddelnano/terraform-provider-mikrotik/client@v0.0.0-00010101000000-000000000000/dns.go:42 +0x10b
github.com/ddelnano/terraform-provider-mikrotik/mikrotik.resourceServerRead({0xc11eb8, 0xc0000bd5c0}, 0xc000159080, {0xb176e0, 0xc000307500})
        github.com/ddelnano/terraform-provider-mikrotik/mikrotik/resource_dns_record.go:61 +0x165
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0003ae000, {0xc11eb8, 0xc0000bd5c0}, 0x24, {0xb176e0, 0xc000307500})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.9.0/helper/schema/resource.go:358 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0003ae000, {0xc11eb8, 0xc0000bd5c0}, 0xc0003404e0, {0xb176e0, 0xc000307500})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.9.0/helper/schema/resource.go:635 +0x35b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc0000b40a8, {0xc11eb8, 0xc0000bd5c0}, 0xc0000bd600)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.9.0/helper/schema/grpc_provider.go:576 +0x534
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc0000baca0, {0xc11f60, 0xc0000f45a0}, 0xc000307860)
        github.com/hashicorp/terraform-plugin-go@v0.4.0/tfprotov5/tf5server/server.go:298 +0x1fb
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0xaf8a40, 0xc0000baca0}, {0xc11f60, 0xc0000f45a0}, 0xc000307800, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.4.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:344 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000dc700, {0xc1db38, 0xc000404300}, 0xc0000c0800, 0xc000400630, 0x108bc70, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1194 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc0000dc700, {0xc1db38, 0xc000404300}, 0xc0000c0800, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1517 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.32.0/server.go:859 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.32.0/server.go:857 +0x294

Error: The terraform-provider-mikrotik_v0.9.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

Anyway, great job with the plugin!

--- a/client/bgp_instance.go
+++ b/client/bgp_instance.go
@@ -127,9 +127,11 @@ func (client Mikrotik) UpdateBgpInstance(b *BgpInstance) (*BgpInstance, error) {
 // DeleteBgpInstance Mikrotik resource
 func (client Mikrotik) DeleteBgpInstance(name string) error {
 	c, err := client.getMikrotikClient()
+	if err != nil {
+		return err
+	}
 
 	bgpInstance, err := client.FindBgpInstance(name)
-
 	if err != nil {
 		return err
 	}

--- a/client/bgp_peer.go
+++ b/client/bgp_peer.go
@@ -117,9 +117,11 @@ func (client Mikrotik) UpdateBgpPeer(b *BgpPeer) (*BgpPeer, error) {
 
 func (client Mikrotik) DeleteBgpPeer(name string) error {
 	c, err := client.getMikrotikClient()
+	if err != nil {
+		return err
+	}
 
 	bgpPeer, err := client.FindBgpPeer(name)
-
 	if err != nil {
 		return err
 	}

--- a/client/client.go
+++ b/client/client.go
@@ -215,6 +215,7 @@ func (client *Mikrotik) getMikrotikClient() (*routeros.Client, error) {
 
 	if err != nil {
 		log.Printf("[ERROR] Failed to login to routerOS with error: %v", err)
+		return nil, err
 	}
 
 	client.connection = mikrotikClient

--- a/mikrotik/resource_bgp_instance.go
+++ b/mikrotik/resource_bgp_instance.go
@@ -126,6 +126,9 @@ func resourceBgpInstanceRead(ctx context.Context, d *schema.ResourceData, m inte
 		d.SetId("")
 		return nil
 	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	return bgpInstanceToData(bgpInstance, d)
 }

--- a/mikrotik/resource_bgp_peer.go
+++ b/mikrotik/resource_bgp_peer.go
@@ -164,6 +164,9 @@ func resourceBgpPeerRead(ctx context.Context, d *schema.ResourceData, m interfac
 		d.SetId("")
 		return nil
 	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	return bgpPeerToData(bgpPeer, d)
 }

--- a/mikrotik/resource_dhcp_lease.go
+++ b/mikrotik/resource_dhcp_lease.go
@@ -68,9 +68,12 @@ func resourceLeaseRead(ctx context.Context, d *schema.ResourceData, m interface{
 
 	lease, err := c.FindDhcpLease(d.Id())
 
-	if err != nil {
+	if _, ok := err.(*client.NotFound); ok {
 		d.SetId("")
 		return nil
+	}
+	if err != nil {
+		return diag.FromErr(err)
 	}
 
 	if lease == nil {

--- a/mikrotik/resource_dns_record.go
+++ b/mikrotik/resource_dns_record.go
@@ -64,6 +64,9 @@ func resourceServerRead(ctx context.Context, d *schema.ResourceData, m interface
 		d.SetId("")
 		return nil
 	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
 	return recordToData(record, d)
 }

--- a/mikrotik/resource_pool.go
+++ b/mikrotik/resource_pool.go
@@ -53,9 +53,12 @@ func resourcePoolRead(ctx context.Context, d *schema.ResourceData, m interface{}
 
 	pool, err := c.FindPool(d.Id())
 
-	if err != nil {
+	if _, ok := err.(*client.NotFound); ok {
 		d.SetId("")
 		return nil
+	}
+	if err != nil {
+		return diag.FromErr(err)
 	}
 
 	return poolToData(pool, d)

--- a/mikrotik/resource_scheduler.go
+++ b/mikrotik/resource_scheduler.go
@@ -61,6 +61,10 @@ func resourceSchedulerRead(ctx context.Context, d *schema.ResourceData, m interf
 	c := m.(*client.Mikrotik)
 
 	scheduler, err := c.FindScheduler(d.Id())
+	if _, ok := err.(*client.NotFound); ok {
+		d.SetId("")
+		return nil
+	}
 	if err != nil {
 		return diag.FromErr(err)
 	}

--- a/mikrotik/resource_script.go
+++ b/mikrotik/resource_script.go
@@ -101,9 +101,12 @@ func resourceScriptRead(ctx context.Context, d *schema.ResourceData, m interface
 
 	script, err := c.FindScript(d.Id())
 
-	if err != nil {
+	if _, ok := err.(*client.NotFound); ok {
 		d.SetId("")
 		return nil
+	}
+	if err != nil {
+		return diag.FromErr(err)
 	}
 
 	return scriptToData(script, d)

